src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs

```diff
@@ -1,207 +1,375 @@
+        }
+
+        internal void StartMultipleMessageListening(CancellationToken cancellationToken)
+        {
+            SessionClient sessionClient = null;
+            IMessageReceiver receiver = null;
+            if (_isSessionsEnabled)
+            {
+                sessionClient = _messagingProvider.CreateSessionClient(_entityPath, _serviceBusAccount.ConnectionString);
+            }
+            else
+            {
+                receiver = Receiver;
+            }
+
+            Task.Run(async () =>
+            {
+                while (true)
+                {
+                    try
+                    {
+                        if (_receiver == null || cancellationToken.IsCancellationRequested)
+                        {
+                            return;
+                        }
+
+                        if (_isSessionsEnabled)
+                        {
+                            try
+                            {
+                                receiver = await sessionClient.AcceptMessageSessionAsync();
+                            }
+                            catch (ServiceBusTimeoutException)
+                            {
+                                // it's expected if the entity is empty, try next time
+                                continue;
+                            }
+                        }
+
+                        IList<Message> messages = messages = await receiver.ReceiveAsync(_serviceBusOptions.BatchOptions.MaxMessageCount, _serviceBusOptions.BatchOptions.OperationTimeout);
+
+                        if (messages != null)
+                        {
+                            Message[] messagesArray = messages.ToArray();
+                            ServiceBusTriggerInput input = new ServiceBusTriggerInput()
+                            {
+                                Messages = messagesArray
+                            };
+                            if (_isSessionsEnabled)
+                            {
+                                input.MessageSession = receiver as IMessageSession;
+                            }
+                            else
+                            {
+                                input.MessageReceiver = receiver as MessageReceiver;
+                            }
+
+                            FunctionResult result = await _triggerExecutor.TryExecuteAsync(GetTriggerFunctionData(messagesArray, input), cancellationToken);
+
+                            if (cancellationToken.IsCancellationRequested)
+                            {
+                                return;
+                            }
+
+                            // Complete batch of messages only if the execution was succesfull
+                            if (result.Succeeded && Receiver != null)
+                            {
+                                await receiver.CompleteAsync(messagesArray.Select(x => x.SystemProperties.LockToken));

```

Comment:
So the contract with the function is that unless it throws an exception, we'll assume all messages were handled successfully and we'll complete them all, otherwise we don't complete any, and they pop back into the queue after their locks expire? How is $dead-letter handled in this case - will the messages be dead lettered after N attempts as normal?

Reply:
Without sessions failed messages automatically go to dead letter queue after N attempts. In sessions case I added the code to abandon messages:
  
Service Bus SDK (`IMessageReceiver.ReceiveAsync`) does nor support `MessageHandlerOptions.AutoComplete` so we do not provide ability to users to complete messages themselves.

Reply:
Right - MessageHandlerOptions.AutoComplete is only used when you're registering a message handler with the receiver. It's assumed that if you're using the receive APIs yourself as you're doing here, that you handle the completions yourself. So we will need to allow users to do that. Why can't we just honor whatever value they've configured for MessageHandlerOptions.AutoComplete and only do the complete/abandon if set to true?